### ISSUE TITLE
Feature request: internationalization / i18n

### ISSUE THREAD

**Issue by Tolc** [vapid/vapid#19]
Simple internationalization would be really nice!

Nothing too fancy, just being able to activate languages site-wise, and translate fields in the dashboard.
Marking fields as translatable could be done by adding a 'i18n' attribute to them in the template, or something like that....

**Comment by asitemade4u** [comment]
+1

**Comment by patrixr** [comment]
Possibly creating per-language tabs on the dashboard pages would be an elegant way of enabling the localisation.

**Comment by gadd-ferreira** [comment]
I am willing to implement it, just need an overview from @srobbin on how he thinks it should be done.

**Comment by srobbin** [comment]
I'm curious to know what you all think about a couple of things:

## Should it be enabled globally, or per-field?

**Globally**

Enabling it globally could be done by adding a config property to `package.json`. e.g.

```
{
  "vapid": {
    "i18n": ["en", "es", "de"]
  }
}
```

Pros: Easy setup, and centralized list of what languages the app supports.
Cons: Need a mechanism for turning it off per-field (e.g. image fields).

**Per-field**

Following the current attribute format, enabling it per-field could look something like this:

```
{{title i18n="en,de,es"}}
{{body type=html i18n="en,de,es"}}
```

Pros: Only enable on certain fields.
Cons: What happens if you want to add another language? Do you have to touch each field again?

## Per-page or per-field tabs

I suppose this depends on the answer to the last question, but does it make more sense to have language tabs per-page, or per-field?

**Per-page**
Pros: Easy for a translator to make contributions.
Cons: How do you handle fields where i18n has been turned off?

**Per-field**
Pros: Easy to identify which fields do/don't have i18n.
Cons: Keep having to click tabs, and might be too easy to enter content into the wrong language field.

---

That said, there might be other ideas and questions. Please chime in.

**Comment by patrixr** [comment]
I feel listing the languages per field is perhaps overkill. I would personally lean towards a global implementation (package.json config looks good), and being able to disable fields, for example like this : 
```
{{ title localised=false }}
```

Personally, I like the idea that the dashboard would reflect exactly what we see on  a page. Meaning that we would see a localised field next to one that doesn't have translations. Per-page tabs feels more inline with that (might mean putting a warning message). 
With per-field tabs, switching a single field to another language would break that consistency across the page.

**Comment by gadd-ferreira** [comment]
I agree with @patrixr,
Global definition, local blacklisting and tab view.

**Comment by asitemade4u** [comment]
Me too!

⁣Stephen Sinclair​

**Comment by storm732** [comment]
Is there any news on this topic?

I am currently working on a website that needs to be available in 2 languages and I would love to use Vapid.

Are there any tricks to implement localization at this point in time? I thought about using two servers with the same code but different databases and redirect requests to the correct one by proxying the two servers through nginx but that would make handling the assets (images) pretty tedious.

**Comment by jakubmirejovsky** [comment]
Hi, any news?

**Comment by Saunved** [comment]
Hi @srobbin 
Is this feature currently being worked on?

**Comment by srobbin** [comment]
@Saunved No, it's not currently being worked on. Are you available to help implement it?

**Comment by Saunved** [comment]
@srobbin Yup, I'd love to help. I went through the comments above. I'll familiarise myself with the source code a bit and keep you updated here once I fork the project and start coding.

**Comment by srobbin** [comment]
@Saunved Thank you! Let me know if you have any questions along the way.

**Comment by Saunved** [comment]
@srobbin I went through parts of the source code. Is there any "developer's guide" or any quickstart for devs that you can lead me to so I can get an overview of how to develop and test Vapid locally? That'd be really useful (for me and others like me). Thanks!
